Button Block: Set proper typography for inner elements

shimotmk opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
The current typography does not work correctly with the given button styles. This fix forces inner elements to inherit the styles.

<!-- In a few words, what is the PR actually doing? -->
Fixes #64662
related #64869

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
1. Set Twenty Twenty-Four theme
2. Place Button block
3. Set the Appearance to Bold, etc.
You can see that the Appearance has changed on either the editor screen or the front screen.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

|Before|After|
|-|-|
|<!-- Before screenshot here -->|<!-- After screenshot here -->|

shimotmk committed:
Fixes #64662

ramonjd commented on packages/block-library/src/button/edit.js:
```diff
@@ -301,6 +317,7 @@ function ButtonEdit( props ) {
 						...colorProps.style,
 						...spacingProps.style,
 						...shadowProps.style,
+						...typographyProps.style,
```
Hmm I think the `.wp-block-button__link` [height of 100%](https://github.com/WordPress/gutenberg/blob/main/packages/block-library/src/button/style.scss#L12-L12) is interfering with the typography setting  `writing-mode: vertical-rl;` 


<img width="243" alt="Screenshot 2024-12-17 at 11 07 42 am" src="https://github.com/user-attachments/assets/eadf61c6-7dcc-4755-8c4e-b58aed5ff4db" />

I wonder if changing it to `auto` would break anything else 🤔

aaronrobertshaw replied:
If `writing-mode` is best placed still on the wrapper, we could just change the `__experimentalSkipSerialization` value for typography to be an array of all the other typography features except writing mode.

ramonjd replied:
Good call! I didn't think of that, thanks!

aaronrobertshaw commented on packages/block-library/src/button/deprecated.js:
```diff
@@ -132,6 +134,188 @@ const blockAttributes = {
 };
 
+const v12 = {
+	attributes: {
```
The attributes here aren't an exact 1:1 match with the [existing `button/block.json`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/button/block.json). I don't think in the context of a deprecation the `role: content` property makes much difference but it's best practice to copy the attributes config as is.

aaronrobertshaw commented on packages/block-library/src/button/deprecated.js:
```diff
@@ -132,6 +134,188 @@ const blockAttributes = {
+		title: {
+			type: 'string',
+			source: 'attribute',
+			selector: 'a',
```
The selector here also doesn't match the [current attribute definition](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/button/block.json#L34) for the Button block.

aaronrobertshaw commented on packages/block-library/src/button/deprecated.js:
```diff
@@ -132,6 +134,188 @@ const blockAttributes = {
+			type: 'string',
+			source: 'html',
+			selector: 'a',
+		},
```
Looks like [`text`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/button/block.json#L38-L42) was changed a year ago to `rich-text`.

aaronrobertshaw commented on packages/block-library/src/button/save.js:
```diff
@@ -38,10 +39,12 @@ export default function save( { attributes, className } ) {
 	const colorProps = getColorClassesAndStyles( attributes );
 	const spacingProps = getSpacingClassesAndStyles( attributes );
 	const shadowProps = getShadowClassesAndStyles( attributes );
+	const typographyProps = getTypographyClassesAndStyles( attributes );
```
@ramonjd given your fluid typography kung fu, can you confirm that we don't need to pass fluid typography settings through here? I'm assuming that that's handled via the `addSaveProps` filter for the typography support.

ramonjd replied:
Can confirm 👍🏻 Good question.

Although it's done via the PHP supports lib:

https://github.com/WordPress/gutenberg/blob/main/lib/block-supports/typography.php#L622-L622

The logic was that blocks should be saved with their raw font sizes, and should only be converted to clamp values if the theme settings say so. 

It guards against the scenario where a user saves a bunch of blocks with custom font sizes that are fluid type, then switches to a theme that doesn't enable fluid typography. The blocks in the database will always have the raw value in the attributes.

aaronrobertshaw commented on test/integration/fixtures/blocks/core__button__deprecated-v12.html:
```diff
@@ -0,0 +1,17 @@
+<!-- wp:buttons {"align":"wide","layout":{"type":"flex","justifyContent":"center"}} -->
+<div class="wp-block-buttons alignwide"><!-- wp:button {"fontSize":"xx-large"} -->
```
The name of this fixture indicates that it targets the `Button` block. The e2e error is due to the top-level block in this fixture being a `Buttons` block. See the full content test for [implementation details](https://github.com/WordPress/gutenberg/blob/trunk/test/integration/full-content/full-content.test.js#L240) of this check.

The best bet here is likely just to make the feature a single Button block with all the typography styles and classes on the wrapper. The easiest way to do this is just checkout trunk and within the editor assign all the various typography styles, then switch to the code editor and copy the block's markup.

aaronrobertshaw reviewed (commented):
Thanks for wrangling this fix @shimotmk 👍 

While reading through the changes I've spotted a few issues and left inline comments for those.

There also looks to be a possible opportunity to clean up the `.has-custom-font-size` class. It appears that was added to allow the inner element to be forced to inherit the parent font size.

Regarding the e2e failures, there are two issues;

1. The new deprecation fixture for the v12 button needs to have a Button block as the base block rather than a Buttons block
2. The v10 fixture has typography styles in it so the final `.serialized.html` result for the fixture needs to have the typography classes and styles moved to the inner element.

The last point there would be cover by running `npm run fixtures:regenerate`. Doing so will also leave a few other unrelated fixtures updated due to the fact their attributes have been reordered. Those changes can be discarded for this PR.

The following diff shows an example of how then fixture could be. It isn't suitable for simply committing to this PR as the fixture should cover all the typography styles (with maybe the exception of writing mode as @ramonjd highlighted).

<details>
<summary>Example diff fixing fixtures</summary>

```diff
diff --git a/test/integration/fixtures/blocks/core__button__deprecated-v10.serialized.html b/test/integration/fixtures/blocks/core__button__deprecated-v10.serialized.html
index e4c7b89c794..0bebe131629 100644
--- a/test/integration/fixtures/blocks/core__button__deprecated-v10.serialized.html
+++ b/test/integration/fixtures/blocks/core__button__deprecated-v10.serialized.html
@@ -1,3 +1,3 @@
 <!-- wp:button {"fontFamily":"cambria-georgia"} -->
-<div class="wp-block-button has-cambria-georgia-font-family"><a class="wp-block-button__link wp-element-button">My button</a></div>
+<div class="wp-block-button"><a class="wp-block-button__link has-cambria-georgia-font-family wp-element-button">My button</a></div>
 <!-- /wp:button -->
diff --git a/test/integration/fixtures/blocks/core__button__deprecated-v12.html b/test/integration/fixtures/blocks/core__button__deprecated-v12.html
index 03ac733e5c3..355f1c2019d 100644
--- a/test/integration/fixtures/blocks/core__button__deprecated-v12.html
+++ b/test/integration/fixtures/blocks/core__button__deprecated-v12.html
@@ -1,17 +1,3 @@
-<!-- wp:buttons {"align":"wide","layout":{"type":"flex","justifyContent":"center"}} -->
-<div class="wp-block-buttons alignwide"><!-- wp:button {"fontSize":"xx-large"} -->
-<div class="wp-block-button has-custom-font-size has-xx-large-font-size"><a class="wp-block-button__link wp-element-button">My button 1</a></div>
-<!-- /wp:button -->
-
-<!-- wp:button {"style":{"typography":{"fontStyle":"normal","fontWeight":"800"}}} -->
-<div class="wp-block-button" style="font-style:normal;font-weight:800"><a class="wp-block-button__link wp-element-button">My button 2</a></div>
-<!-- /wp:button -->
-
-<!-- wp:button {"style":{"typography":{"letterSpacing":"39px"}}} -->
-<div class="wp-block-button" style="letter-spacing:39px"><a class="wp-block-button__link wp-element-button">My button 3</a></div>
-<!-- /wp:button -->
-
-<!-- wp:button {"tagName":"button","style":{"typography":{"letterSpacing":"39px"}}} -->
-<div class="wp-block-button" style="letter-spacing:39px"><button type="button" class="wp-block-button__link wp-element-button">My button 4</button></div>
-<!-- /wp:button --></div>
-<!-- /wp:buttons -->
+<!-- wp:button {"fontSize":"xx-large","style":{"typography":{"fontStyle":"normal","fontWeight":"800","letterSpacing":"39px"}}} -->
+<div class="wp-block-button has-custom-font-size has-xx-large-font-size" style="font-style:normal;font-weight:800;letter-spacing:39px"><a class="wp-block-button__link wp-element-button">My Button</a></div>
+<!-- /wp:button -->
\ No newline at end of file
diff --git a/test/integration/fixtures/blocks/core__button__deprecated-v12.json b/test/integration/fixtures/blocks/core__button__deprecated-v12.json
index b6cab91f0d5..7fd537b6575 100644
--- a/test/integration/fixtures/blocks/core__button__deprecated-v12.json
+++ b/test/integration/fixtures/blocks/core__button__deprecated-v12.json
@@ -1,72 +1,20 @@
 [
 	{
-		"name": "core/buttons",
+		"name": "core/button",
 		"isValid": true,
 		"attributes": {
-			"align": "wide",
-			"layout": {
-				"type": "flex",
-				"justifyContent": "center"
+			"tagName": "a",
+			"type": "button",
+			"text": "My Button",
+			"fontSize": "xx-large",
+			"style": {
+				"typography": {
+					"fontStyle": "normal",
+					"fontWeight": "800",
+					"letterSpacing": "39px"
+				}
 			}
 		},
-		"innerBlocks": [
-			{
-				"name": "core/button",
-				"isValid": true,
-				"attributes": {
-					"tagName": "a",
-					"type": "button",
-					"text": "My button 1",
-					"fontSize": "xx-large"
-				},
-				"innerBlocks": []
-			},
-			{
-				"name": "core/button",
-				"isValid": true,
-				"attributes": {
-					"tagName": "a",
-					"type": "button",
-					"text": "My button 2",
-					"style": {
-						"typography": {
-							"fontStyle": "normal",
-							"fontWeight": "800"
-						}
-					}
-				},
-				"innerBlocks": []
-			},
-			{
-				"name": "core/button",
-				"isValid": true,
-				"attributes": {
-					"tagName": "a",
-					"type": "button",
-					"text": "My button 3",
-					"style": {
-						"typography": {
-							"letterSpacing": "39px"
-						}
-					}
-				},
-				"innerBlocks": []
-			},
-			{
-				"name": "core/button",
-				"isValid": false,
-				"attributes": {
-					"tagName": "button",
-					"type": "button",
-					"text": "My button 4",
-					"style": {
-						"typography": {
-							"letterSpacing": "39px"
-						}
-					}
-				},
-				"innerBlocks": []
-			}
-		]
+		"innerBlocks": []
 	}
 ]
diff --git a/test/integration/fixtures/blocks/core__button__deprecated-v12.parsed.json b/test/integration/fixtures/blocks/core__button__deprecated-v12.parsed.json
index 466b53ae00e..542817c8428 100644
--- a/test/integration/fixtures/blocks/core__button__deprecated-v12.parsed.json
+++ b/test/integration/fixtures/blocks/core__button__deprecated-v12.parsed.json
@@ -1,84 +1,20 @@
 [
 	{
-		"blockName": "core/buttons",
+		"blockName": "core/button",
 		"attrs": {
-			"align": "wide",
-			"layout": {
-				"type": "flex",
-				"justifyContent": "center"
+			"fontSize": "xx-large",
+			"style": {
+				"typography": {
+					"fontStyle": "normal",
+					"fontWeight": "800",
+					"letterSpacing": "39px"
+				}
 			}
 		},
-		"innerBlocks": [
-			{
-				"blockName": "core/button",
-				"attrs": {
-					"fontSize": "xx-large"
-				},
-				"innerBlocks": [],
-				"innerHTML": "\n<div class=\"wp-block-button has-custom-font-size has-xx-large-font-size\"><a class=\"wp-block-button__link wp-element-button\">My button 1</a></div>\n",
-				"innerContent": [
-					"\n<div class=\"wp-block-button has-custom-font-size has-xx-large-font-size\"><a class=\"wp-block-button__link wp-element-button\">My button 1</a></div>\n"
-				]
-			},
-			{
-				"blockName": "core/button",
-				"attrs": {
-					"style": {
-						"typography": {
-							"fontStyle": "normal",
-							"fontWeight": "800"
-						}
-					}
-				},
-				"innerBlocks": [],
-				"innerHTML": "\n<div class=\"wp-block-button\" style=\"font-style:normal;font-weight:800\"><a class=\"wp-block-button__link wp-element-button\">My button 2</a></div>\n",
-				"innerContent": [
-					"\n<div class=\"wp-block-button\" style=\"font-style:normal;font-weight:800\"><a class=\"wp-block-button__link wp-element-button\">My button 2</a></div>\n"
-				]
-			},
-			{
-				"blockName": "core/button",
-				"attrs": {
-					"style": {
-						"typography": {
-							"letterSpacing": "39px"
-						}
-					}
-				},
-				"innerBlocks": [],
-				"innerHTML": "\n<div class=\"wp-block-button\" style=\"letter-spacing:39px\"><a class=\"wp-block-button__link wp-element-button\">My button 3</a></div>\n",
-				"innerContent": [
-					"\n<div class=\"wp-block-button\" style=\"letter-spacing:39px\"><a class=\"wp-block-button__link wp-element-button\">My button 3</a></div>\n"
-				]
-			},
-			{
-				"blockName": "core/button",
-				"attrs": {
-					"tagName": "button",
-					"style": {
-						"typography": {
-							"letterSpacing": "39px"
-						}
-					}
-				},
-				"innerBlocks": [],
-				"innerHTML": "\n<div class=\"wp-block-button\" style=\"letter-spacing:39px\"><button type=\"button\" class=\"wp-block-button__link wp-element-button\">My button 4</button></div>\n",
-				"innerContent": [
-					"\n<div class=\"wp-block-button\" style=\"letter-spacing:39px\"><button type=\"button\" class=\"wp-block-button__link wp-element-button\">My button 4</button></div>\n"
-				]
-			}
-		],
-		"innerHTML": "\n<div class=\"wp-block-buttons alignwide\">\n\n\n\n\n\n</div>\n",
+		"innerBlocks": [],
+		"innerHTML": "\n<div class=\"wp-block-button has-custom-font-size has-xx-large-font-size\" style=\"font-style:normal;font-weight:800;letter-spacing:39px\"><a class=\"wp-block-button__link wp-element-button\">My Button</a></div>\n",
 		"innerContent": [
-			"\n<div class=\"wp-block-buttons alignwide\">",
-			null,
-			"\n\n",
-			null,
-			"\n\n",
-			null,
-			"\n\n",
-			null,
-			"</div>\n"
+			"\n<div class=\"wp-block-button has-custom-font-size has-xx-large-font-size\" style=\"font-style:normal;font-weight:800;letter-spacing:39px\"><a class=\"wp-block-button__link wp-element-button\">My Button</a></div>\n"
 		]
 	}
 ]
diff --git a/test/integration/fixtures/blocks/core__button__deprecated-v12.serialized.html b/test/integration/fixtures/blocks/core__button__deprecated-v12.serialized.html
index 2df6789245e..1b18b70facc 100644
--- a/test/integration/fixtures/blocks/core__button__deprecated-v12.serialized.html
+++ b/test/integration/fixtures/blocks/core__button__deprecated-v12.serialized.html
@@ -1,17 +1,3 @@
-<!-- wp:buttons {"align":"wide","layout":{"type":"flex","justifyContent":"center"}} -->
-<div class="wp-block-buttons alignwide"><!-- wp:button {"fontSize":"xx-large"} -->
-<div class="wp-block-button has-custom-font-size"><a class="wp-block-button__link has-xx-large-font-size wp-element-button">My button 1</a></div>
+<!-- wp:button {"fontSize":"xx-large","style":{"typography":{"fontStyle":"normal","fontWeight":"800","letterSpacing":"39px"}}} -->
+<div class="wp-block-button has-custom-font-size"><a class="wp-block-button__link has-xx-large-font-size wp-element-button" style="font-style:normal;font-weight:800;letter-spacing:39px">My Button</a></div>
 <!-- /wp:button -->
-
-<!-- wp:button {"style":{"typography":{"fontStyle":"normal","fontWeight":"800"}}} -->
-<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" style="font-style:normal;font-weight:800">My button 2</a></div>
-<!-- /wp:button -->
-
-<!-- wp:button {"style":{"typography":{"letterSpacing":"39px"}}} -->
-<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" style="letter-spacing:39px">My button 3</a></div>
-<!-- /wp:button -->
-
-<!-- wp:button {"tagName":"button","style":{"typography":{"letterSpacing":"39px"}}} -->
-<div class="wp-block-button" style="letter-spacing:39px"><button type="button" class="wp-block-button__link wp-element-button">My button 4</button></div>
-<!-- /wp:button --></div>
-<!-- /wp:buttons -->

```
</details>

I hope that helps 🤞

shimotmk committed:
fix

shimotmk commented:
@ramonjd  @aaronrobertshaw
Thanks for reviewing!

I adjusted the attributes and made core__button__deprecated-v12.html into a single button block.

However, the data for the fourth letterSpacing setting in core__button__deprecated-v12.html cannot be migrated.

Why is that?🤔

aaronrobertshaw commented:
> However, the data for the fourth letterSpacing setting in core__button__deprecated-v12.html cannot be migrated.

I'm not following this comment sorry. It looks fine to me in the [lastest commit](https://github.com/WordPress/gutenberg/pull/68023/commits/aaaa4522918e7ad198072c54a7092750c76f9327#diff-9fd17ee39d7199846830f221d889844977ba0e973170845a51f1cd3fcf86f2f7R14).

The current e2e failures relate to `full post content fixture › core__button__deprecated-v10`. As I noted in my [review](https://github.com/WordPress/gutenberg/pull/68023#pullrequestreview-2507670219) this fixture needing updating. The required change was also in the diff I provided. 

After updating the v10 fixture, all the fixture tests pass for me.

Once that is sorted, there's still the issue with writing mode to fix up.

shimotmk commented:
@aaronrobertshaw 
Sorry. I was mistaken. The transition was going well.

shimotmk committed:
fix writing-mode
https://github.com/WordPress/gutenberg/pull/68023#discussion_r1887732639

shimotmk commented:
@ramonjd @aaronrobertshaw
I adjusted the writing-mode to add inline styles to the outer div 🙂

ramonjd commented on packages/block-library/src/button/edit.js:
```diff
@@ -301,6 +317,8 @@ function ButtonEdit( props ) {
 						...colorProps.style,
 						...spacingProps.style,
 						...shadowProps.style,
+						...typographyProps.style,
```
It looks like `blockProps` is still adding the style attributes to the wrapper element.

I'm not sure if this is the perfect way to go about it, but we might have to explicitly exclude these keys in the hook arguments above. For example:

```js
	const blockProps = useBlockProps( {
		ref: useMergeRefs( [ setPopoverAnchor, ref ] ),
		onKeyDown,
		style: {
			fontSize: undefined,
			lineHeight: undefined,
			fontFamily: undefined,
			fontWeight: undefined,
			fontStyle: undefined,
			textTransform: undefined,
			textDecoration: undefined,
			letterSpacing: undefined,
		},
	} );
```

aaronrobertshaw replied:
> It looks like blockProps is still adding the style attributes to the wrapper element.

This might suggest a regression around the skip serialization feature. I'll try and dig a bit deeper later today or tomorrow after I push the block support stabilization changes forward.

aaronrobertshaw replied:
I've had a quick look at that latest changes and it appears we just need to tweak the properties listed within the `__experimentalSkipSerialization` array. That array doesn't use the `__experimental` prefix for properties. 

If the history here interests, you can find the original PR that allowed skipping individual features in https://github.com/WordPress/gutenberg/pull/36293. An existing example of skipping an experimental feature is the Navigation block and its `textDecoration` even though the support flag for that is still experimental.

shimotmk replied:
Thanks! I fixed it 🐛

ramonjd commented on packages/block-library/src/button/block.json:
```diff
@@ -85,6 +85,16 @@
 			}
 		},
 		"typography": {
+			"__experimentalSkipSerialization": [
+				"fontSize",
+				"lineHeight",
+				"__experimentalFontFamily",
+				"__experimentalFontWeight",
+				"__experimentalFontStyle",
+				"__experimentalTextTransform",
+				"__experimentalTextDecoration",
+				"__experimentalLetterSpacing"
+			],
```
Do you think it's okay remove the `__experimental` prefix here and use the [stable keys](https://github.com/WordPress/gutenberg/pull/63401) @aaronrobertshaw?

aaronrobertshaw replied:
I did wonder the same thing but given that the stabilization is in a state of flux right now it might be better to hold off on using the stable keys. 

There's a good chance I'll have to update them all as part of the latest approach to block support stabilization. So a few more keys are neither here nor there.

shimotmk committed:
fix https://github.com/WordPress/gutenberg/pull/68023#discussion_r1889224581

shimotmk commented:
@ramonjd @aaronrobertshaw 
Any other things I need to fix?

aaronrobertshaw commented:
> Any other things I need to fix?

@shimotmk the issue I flagged in my last two reviews still needs fixing.

> The current e2e failures relate to full post content fixture › core__button__deprecated-v10. As I noted in my https://github.com/WordPress/gutenberg/pull/68023#pullrequestreview-2507670219 this fixture needing updating. The required change was also in the diff I provided.

shimotmk committed:
fix
https://github.com/WordPress/gutenberg/pull/68023#pullrequestreview-2507670219

shimotmk commented:
@aaronrobertshaw
Thanks for the reply

Moved the `has-custom-font-size` class internally.
And updated `test/integration/fixtures/blocks/core__button__deprecated-v10.serialized.html`
Is this OK?

ramonjd commented:
Thanks for the continued work on this PR @shimotmk 🙇🏻 

It looks like the mobile unit tests fails are related to the button block: https://github.com/WordPress/gutenberg/actions/runs/12429816613/job/34704026313?pr=68023#step:6:2906

shimotmk commented:
@ramonjd
Thanks for the reply
I don't know how to fix the mobile unit.
How do I fix it?

ramonjd commented:
I think the hook needs to be exported for native:

```diff
diff --git a/packages/block-editor/src/hooks/index.native.js b/packages/block-editor/src/hooks/index.native.js
index c7f9df868f..0e4c2aa276 100644
--- a/packages/block-editor/src/hooks/index.native.js
+++ b/packages/block-editor/src/hooks/index.native.js
@@ -33,3 +33,4 @@ export { getColorClassesAndStyles, useColorProps } from './use-color-props';
 export { getSpacingClassesAndStyles } from './use-spacing-props';
 export { useCachedTruthy } from './use-cached-truthy';
 export { useEditorWrapperStyles } from './use-editor-wrapper-styles';
+export { getTypographyClassesAndStyles } from './use-typography-props';
```

Try that, see if it works! 😄

shimotmk committed:
try
https://github.com/WordPress/gutenberg/pull/68023#issuecomment-2558602641

shimotmk commented:
@ramonjd
Thank you! 
It looks like the test passed🙌.

ramonjd commented:
This is nearly ready, thank you for all the work here @shimotmk 

There's one thing I noticed with global styles. If you add a writing mode value to the button block in global styles, it's applied to the inner `.wp-block-button .wp-block-button__link`, which was a similar issue to [above](https://github.com/WordPress/gutenberg/pull/68023#discussion_r1887732639).

<img width="750" alt="Screenshot 2024-12-24 at 7 56 26 am" src="https://github.com/user-attachments/assets/515d97f3-5d09-4495-8a04-87d34f6ca444" />

I'm not 100% sure of the remedy. I tried converting `button/block.json` to use the [selectors API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-selectors/):


```json
	"selectors": {
		"root": ".wp-block-button .wp-block-button__link",
		"typography": {
			"writing-mode": ".wp-block-button"
		}
	}
```

But I think I'm missing something as it doesn't work.

If you don't mind, we might wait to ask @aaronrobertshaw for his opinion, as he knows this stuff better than I.

ramonjd commented:
> I tried converting button/block.json to use the [selectors API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-selectors/):

It should be

```json
	"selectors": {
		"root": ".wp-block-button .wp-block-button__link",
		"typography": {
			"writingMode": ".wp-block-button"
		}
	}
```

And it should work, but I think it's being filtered out via `wp_theme_json_get_style_nodes` here:

https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json-gutenberg.php#L2667-L2667

The filter is added here: 

https://github.com/WordPress/gutenberg/blob/trunk/lib/script-loader.php#L43

If I comment out that line it works!

So maybe something to do with [wp_filter_out_block_nodes](https://github.com/WordPress/wordpress-develop/blob/779ed48a034c60a65cd1ac7ab35262e496e39c99/src/wp-includes/script-loader.php#L2475) - I haven't tested that deeply yet.

ramonjd commented:
> So maybe something to do with [wp_filter_out_block_nodes](https://github.com/WordPress/wordpress-develop/blob/779ed48a034c60a65cd1ac7ab35262e496e39c99/src/wp-includes/script-loader.php#L2475) - I haven't tested that deeply yet.

Nope 😄 It was because of the [block style sheet caching](https://github.com/WordPress/gutenberg/blob/main/lib/global-styles-and-settings.php#L261-L261). I cleared my cache and it was 👍🏻 

I still think we need to add the following to the block.json file so that the writing mode in global styles is added to the wrapper:

```diff
diff --git a/packages/block-library/src/button/block.json b/packages/block-library/src/button/block.json
index 953fcae023..6fcb7aca4c 100644
--- a/packages/block-library/src/button/block.json
+++ b/packages/block-library/src/button/block.json
@@ -132,7 +132,6 @@
 				"width": true
 			}
 		},
-		"__experimentalSelector": ".wp-block-button .wp-block-button__link",
 		"interactivity": {
 			"clientNavigation": true
 		}
@@ -142,5 +141,11 @@
 		{ "name": "outline", "label": "Outline" }
 	],
 	"editorStyle": "wp-block-button-editor",
-	"style": "wp-block-button"
+	"style": "wp-block-button",
+	"selectors": {
+		"root": ".wp-block-button .wp-block-button__link",
+		"typography": {
+			"writingMode": ".wp-block-button"
+		}
+	}
 }
```

Then I think this PR is good to go. 

Thank you!!

shimotmk committed:
fix https://github.com/WordPress/gutenberg/pull/68023#issuecomment-2565003954

shimotmk commented:
@ramonjd

Thanks
It was working fine, so I fixed it with the fix

> I still think we need to add the following to the block.json file so that the writing mode in global styles is added to the wrapper:
> 
> ```diff
> diff --git a/packages/block-library/src/button/block.json b/packages/block-library/src/button/block.json
> index 953fcae023..6fcb7aca4c 100644
> --- a/packages/block-library/src/button/block.json
> +++ b/packages/block-library/src/button/block.json
> @@ -132,7 +132,6 @@
>  				"width": true
>  			}
>  		},
> -		"__experimentalSelector": ".wp-block-button .wp-block-button__link",
>  		"interactivity": {
>  			"clientNavigation": true
>  		}
> @@ -142,5 +141,11 @@
>  		{ "name": "outline", "label": "Outline" }
>  	],
>  	"editorStyle": "wp-block-button-editor",
> -	"style": "wp-block-button"
> +	"style": "wp-block-button",
> +	"selectors": {
> +		"root": ".wp-block-button .wp-block-button__link",
> +		"typography": {
> +			"writingMode": ".wp-block-button"
> +		}
> +	}
>  }
> ```

ramonjd reviewed (approved):
I tested:

✅  Settings styles for `style.block['core/button']` in theme.json works as expected
✅  Global styles are applied correctly, with writing mode set on the block wrapper
✅  Block-level styles override global styles


Thank you for all the work on this PR @shimotmk 🙇🏻

shimotmk commented:
@ramonjd
Thanks for the review!
